relayer/pkg/starknet/client.go

```diff
@@ -11,69 +13,99 @@ import (
+	// provider interface
+	BlockByHash(context.Context, string, string) (*caigotypes.Block, error)
+	BlockByNumber(context.Context, *big.Int, string) (*caigotypes.Block, error)
+	Call(context.Context, caigotypes.FunctionCall, string) ([]string, error)
+	ChainID(context.Context) (string, error)
 }
 
 type Writer interface {
+	AccountNonce(context.Context, string) (*big.Int, error)
+	Invoke(context.Context, caigotypes.Transaction) (*caigotypes.AddTxResponse, error)
+	TransactionByHash(context.Context, string) (*caigotypes.Transaction, error)
+	TransactionReceipt(context.Context, string) (*caigotypes.TransactionReceipt, error)
+	EstimateFee(context.Context, caigotypes.Transaction) (*caigotypes.FeeEstimate, error)
+}
+
+type Unimplemented interface {
+	Class(context.Context, string) (*caigotypes.ContractClass, error)
+	ClassHashAt(context.Context, string) (*caigotypes.Felt, error)
+	ClassAt(context.Context, string) (*caigotypes.ContractClass, error)
 }
 
 type ReaderWriter interface {
 	Reader
 	Writer
+	Unimplemented
 }
 
 var _ ReaderWriter = (*Client)(nil)
 
+var _ caigotypes.Provider = (*Client)(nil)
+
 type Client struct {
-	gw   *caigogw.Gateway
+	gw   *caigogw.GatewayProvider
 	lggr logger.Logger
+	cfg  Config
 }
 
-func NewClient(chainID string, lggr logger.Logger) (*Client, error) {
-	return &Client{
-		gw:   caigogw.NewClient(caigogw.WithChain(chainID)),
+func NewClient(chainID string, baseURL string, lggr logger.Logger, cfg Config) (*Client, error) {
+	client := &Client{
+		gw:   caigogw.NewProvider(caigogw.WithChain(chainID)),
 		lggr: lggr,
-	}, nil
+		cfg:  cfg,
+	}
+	client.setURL(baseURL) // hack: change the base URL (not supported in caigo)
+
+	return client, nil
 }
 
+func (c *Client) setURL(baseURL string) {
+	if baseURL == "" {
+		return // if empty, use default from caigo
+	}
+
+	c.gw.Gateway.Base = baseURL
+	c.gw.Gateway.Feeder = baseURL + "/feeder_gateway"
+	c.gw.Gateway.Gateway = baseURL + "/gateway"
+}
+
+// -- Custom Wrapped Func --
+
 func (c *Client) CallContract(ctx context.Context, ops CallOps) (res []string, err error) {
-	tx := caigotypes.Transaction{
+	tx := caigotypes.FunctionCall{
 		ContractAddress:    ops.ContractAddress,
 		EntryPointSelector: ops.Selector,
 		Calldata:           ops.Calldata,
 	}
 
-	res, err = c.gw.Call(ctx, tx, nil)
+	res, err = c.Call(ctx, tx, "")
 	if err != nil {
-		return res, errors.Wrap(err, "couldn't call the contract")
+		return res, errors.Wrap(err, "error in client.CallContract")
 	}
 
 	return
 }
 
-func (c *Client) ChainID(ctx context.Context) (id string, err error) {
-	id, err = c.gw.ChainID(ctx)
-	if err != nil {
-		return id, errors.Wrap(err, "couldn't get chain id")
-	}
-
-	return id, nil
-}
-
-func (c *Client) LatestBlockHeight(ctx context.Context) (height uint64, err error) {
+func (c *Client) LatestBlockHeight(parentCtx context.Context) (height uint64, err error) {
+	ctx, cancel := context.WithTimeout(parentCtx, c.cfg.RequestTimeout())
```

Comment:
i think that would also work! 

personally, I like having the client with a built in timeout that way there is one place (`cfg`) where the timeout is defined for any query. also it helps because you don't have to worry about defining any query timeouts when you use the client - and you can pass other contexts directly in

Reply:
My opinion on that is that the client is a utility module and it's behaviour is to be defined by the consumer (such as contract_reader) by passing the context in. It also removes the dependency on cfg that is imo unnecessary here. Definitely we don't want to change the code every time, I just mean that the timeout is to be defined upstream (in contract_reader or even earlier because contract_reader also receives the context from its consumer).

Reply:
that's a valid point!

my main concern is if the upstream implementation doesn't include a timeout, then the request has the potential to hang and block execution. so by having a parameter set on a lower level of the client it reduces the chance of that happening?

if you feel strongly about it, I'm happy to change it :) we also specify a client level timeout in the solana and terra implementations

Reply:
I'm totally fine with you giving it another thought and deciding it's fine this way :)

I noticed that our modules are tightly coupled (i.e. every stores its copy of `cfg`) and it's hard to move them around. I think we should continuously address it since the app is growing. On example of this client, if let's say another module would require to use it (on Solana and Terra there was a getter for it in the `Chain` interface, so the core module can potentially call it), that module would have to comply with this timeout although it didn't set it. Or if this module sets the timeout itself through the context, there could be unexpected results.